File: string/917.py
class Solution:
    def reverseOnlyLetters(self, S: str) -> str:
        pre = 0
        nxt = len(S) - 1
        
        prelist = []
        nxtlist = []
        
        def is_good(c):
            return c.isalpha()
            # return 65 <= ord(c) <= 90 or 97 <= ord(c) <= 122
        
        while pre < nxt:
            
            while pre < nxt and not is_good(S[pre]):
                prelist.append(S[pre])
                pre += 1
                
            while pre < nxt and not is_good(S[nxt]):
                nxtlist.append(S[nxt])
                nxt -= 1
                
            if pre >= nxt:
                break
            prelist.append(S[nxt])
            nxt -= 1
            nxtlist.append(S[pre])
            pre += 1
            
        if pre == nxt:
            prelist.append(S[pre])
        nxtlist.reverse()
        return ''.join(prelist)+''.join(nxtlist)
            

File: two_pointers/917.cc
class Solution {
public:
    string reverseOnlyLetters(string s) {
        for (int i = 0, j = s.size()-1; i < j;) {
            while (i < j && (s[i] < 65 || s[i] > 90 && s[i] < 97))
                ++i;
            while (i < j && (s[j] < 65 || s[j] > 90 && s[j] < 97))
                --j;
            char temp = s[i];
            s[i] = s[j];
            s[j] = temp;
            ++i;
            --j;
        }
        return s;
    }
};
